### ISSUE TITLE
[BUG]destbufferoverflow in pair_bop.cpp:5308:22

### ISSUE THREAD

**Issue by junghans** [lammps/lammps#2078]
openSUSE found the following error:
```
[  536s]     inlined from 'void LAMMPS_NS::PairBOP::read_table(char*)' at /home/abuild/rpmbuild/BUILD/lammps-patch_5May2020/src/MANYBODY/pair_bop.cpp:5308:22:
[  536s] /usr/include/bits/stdio2.h:260:58: warning: call to '__fgets_chk_warn' declared with attribute warning: fgets called with bigger size than length of destination buffer
[  536s]   return __fgets_chk_warn (__s, __bos (__s), __n, __stream);
[  536s]                                                           ^
```

**Comment by junghans** [comment]
Actually it had a couple more warnings, too, but that was the only error:
```
[ 2201s] 
[ 2201s] I: Program causes undefined operation
[ 2201s]    (likely same variable used twice and post/pre incremented in the same expression).
[ 2201s]    e.g. x = x++; Split it in two operations.
[ 2201s] W: lammps sequence-point /home/abuild/rpmbuild/BUILD/lammps-patch_5May2020/src/KOKKOS/fix_qeq_reax_kokkos.cpp:56
[ 2201s] W: lammps sequence-point /home/abuild/rpmbuild/BUILD/lammps-patch_5May2020/src/USER-MGPT/pair_mgpt.cpp:2148, 2149, 2150, 2151, 2167, 2168, 2169, 2170
[ 2201s] 
[ 2201s] I: Statement might potentially overflow a destination buffer, where a size larger
[ 2201s]     than the actual buffer was specified
[ 2201s] E: lammps destbufferoverflow /home/abuild/rpmbuild/BUILD/lammps-patch_5May2020/src/MANYBODY/pair_bop.cpp:5308:22
```

**Comment by akohlmey** [comment]
> openSUSE found the following error:
> 
> ```
> [  536s]     inlined from 'void LAMMPS_NS::PairBOP::read_table(char*)' at /home/abuild/rpmbuild/BUILD/lammps-patch_5May2020/src/MANYBODY/pair_bop.cpp:5308:22:
> [  536s] /usr/include/bits/stdio2.h:260:58: warning: call to '__fgets_chk_warn' declared with attribute warning: fgets called with bigger size than length of destination buffer
> [  536s]   return __fgets_chk_warn (__s, __bos (__s), __n, __stream);
> [  536s]                                                           ^
> ```

This looks like a false positive.  The call at line 5308 is:
```
         while(fgets(s,MAXLINE,fp)!=NULL&&pass==0) {
```
and the array `s` is dimensioned to 'MAXLINE' in line 5026. fgets reads at most one byte less than MAXLINE and appends a '\0`, so I don't see where this is writing beyond the buffer.

**Comment by junghans** [comment]
It is gcc-7.4.0, build log [here](https://github.com/lammps/lammps/files/4605070/x86_64_build.txt)

**Comment by junghans** [comment]
gcc-9.3.1 has the same warning.

**Comment by akohlmey** [comment]
This is not a bug in the code, but it is due to openSUSE's additional compiler settings. The warning goes away, if you change
``` c++
void _noopt PairBOP::read_table(char *filename)
```
to
``` c++
void PairBOP::read_table(char *filename)
```
The reason is that using `-D_FORTIFY_SOURCE` requires compiling with optimization, but that macro turns optimization and variable tracking off for this function. If you turn off optimization globally, you should see a warning from `/usr/include/features.h` about that requirement.

**Comment by akohlmey** [comment]
Here is a short demonstration. It print the warning for only one function as soon as you compile with `-O` and warn about enabling optimization without.
``` c++
#define _FORTIFY_SOURCE 2

#include <cstdio>
#define MAXLINE 1024

void __attribute__((optimize("O0"))) readwarn(FILE *fp)
{
    char s[MAXLINE];

    while(fgets(s,MAXLINE,fp))
        printf("read characters for line: %s",s);
}

void readnowarn(FILE *fp)
{
    char s[MAXLINE];

    while(fgets(s,MAXLINE,fp))
        printf("read characters for line: %s",s);
}

int main(int, char **)
{
    FILE *fp = fopen("some_file","r");
    if (fp) {
        readwarn(fp);
        readnowarn(fp);
    }
    fclose(fp);
    return 0;
}
```

**Comment by junghans** [comment]
Thanks for the explaination!

One last question, why are we having `_noopt` on `PairBOP::read_table()`?

**Comment by akohlmey** [comment]
> Thanks for the explaination!
> 
> One last question, why are we having `_noopt` on `PairBOP::read_table()`?

Because this (and several other functions) put a undue strain on compilers with high optimization and would cause confusing warnings and needless delays in compilation, since there is nothing there that needs optimizing.

**Comment by junghans** [comment]
I added `sed -i 's/_noopt//' src/KSPACE/pppm_disp.cpp src/MANYBODY/pair_bop.cpp src/force.cpp src/lammps.cpp src/modify.cpp` to the openSUSE spec file and gcc seems to eat it without problem from SLE 15 SP2 over leap 15.[12] to Tumbleweed.

Can we add an option to CMake that is off by default to disable `_noopt`?

**Comment by junghans** [comment]
I like afd769314b67d9614e70d80c4e8fade41b367899

**Comment by junghans** [comment]
On a second thought, I would still like a mechanism that allows to disable `_noopt` completely independently if `-D_FORTIFY_SOURCE` is set or not.

**Comment by akohlmey** [comment]
> On a second thought, I would still like a mechanism that allows to disable `_noopt` completely independently if `-D_FORTIFY_SOURCE` is set or not.

but i don't.
what is implemented in PR #2080 is as far as i am willing to go to accommodate a flaw outside of LAMMPS.

**Comment by junghans** [comment]
> > On a second thought, I would still like a mechanism that allows to disable `_noopt` completely independently if `-D_FORTIFY_SOURCE` is set or not.
> 
> but i don't.
> what is implemented in PR #2080 is as far as i am willing to go to accommodate a flaw outside of LAMMPS.

Your call!

Have you seen any significant compile time degradation using a recent gcc like 9 or 10 on the 6 places where we use `_noopt`?

**Comment by akohlmey** [comment]
Compilation time is only one of the issues. As far as I am concerned, there is nothing more to say on this topic.
